File: 21.py
class TreeNode:
    def __init__(self, key=None, left=None, right=None):
        self.key = key
        self.left = left
        self.right = right
        self.l_height = -1
        self.r_height = -1
        self.l_size = 0
        self.r_size = 0

    def __iter__(self):
        if self:
            if self.left:
                for node in self.left:
                    yield node
            yield self
            if self.right:
                for node in self.right:
                    yield node

    def _count_subtrees_(self):
        if self.left:
            self.l_size = len([i for i in self.left])
        if self.right:
            self.r_size = len([i for i in self.right])


class BinarySearchTree:
    def __init__(self, root=None):
        self.root = root

    def __iter__(self):
        return self.root.__iter__()

    def get_root(self):
        return self.root

    def put(self, key):
        if self.root is None:
            self.root = TreeNode(key)
        else:
            parent = None
            node = self.root
            while node is not None:
                parent = node
                if key < node.key:
                    node = node.left
                elif key > node.key:
                    node = node.right
                else:
                    return

            if key < parent.key:
                parent.left = TreeNode(key)
            elif key > parent.key:
                parent.right = TreeNode(key)

    def minimum(self, node):
        current = node
        while current.left is not None:
            current = current.left
        return current

    def remove(self, key):
        self.root = self._remove_(self.root, key)

    def _remove_(self, root, key):
        if root is None:
            return root
        if key < root.key:
            root.left = self._remove_(root.left, key)
        elif key > root.key:
            root.right = self._remove_(root.right, key)
        elif root.left and root.right:
            root.key = self.minimum(root.right).key
            root.right = self._remove_(root.right, root.key)
        elif root.left:
            root = root.left
        else:
            root = root.right
        return root

    def _height_(self, node):
        if not node:
            return 0
        if node.left:
            l_h = self._height_(node.left)
        else:
            l_h = -1
        if node.right:
            r_h = self._height_(node.right)
        else:
            r_h = -1
        return max(l_h, r_h) + 1

    def update_info(self):
        for node in self:
            node.l_height = self._height_(node.left)
            node.r_height = self._height_(node.right)
        for node in self:
            node._count_subtrees_()

    def preorder(self, node):
        if node:
            print(node.key)
            self.preorder(node.left)
            self.preorder(node.right)

    def preorder_file(self, node, file):
        if node:
            file.write(str(node.key) + "\n")
            self.preorder_file(node.left, file)
            self.preorder_file(node.right, file)

if __name__ == '__main__':
    bst = BinarySearchTree()
    with open("in.txt", "r") as fin:
        for key in fin.readlines():
            bst.put(int(key))
    bst.update_info()

    special_keys = []
    for node in bst:
        if node.left and node.right:
            if node.l_height == node.r_height and node.l_size != node.r_size:
                special_keys.append(node.key)

    if len(special_keys) % 2 != 0:
        special_keys.sort()
        key_to_remove = special_keys[len(special_keys) // 2]
        bst.remove(key_to_remove)

    with open("out.txt", "w") as fout:
        bst.preorder_file(bst.get_root(), fout)

File: 21.cpp
#include <iostream>
#include <fstream>
#include <algorithm>
#include <vector>

class BinaryTree {
	private:
	    struct Node {
		    int Info;
	        Node* LSon;
	        Node* RSon;
            int LHeight;
            int RHeight;
            int LSize;
            int RSize;

            void PreOrderCount(Node*, std::vector<int>&) const;
            void CountSubtrees(Node*);
	    };
	    Node* Root;
	public:
	    BinaryTree();
	    ~BinaryTree();
	    BinaryTree::Node* Insert(Node*, const int&);
	    BinaryTree::Node* EraseByValue(Node*, int&);
	    BinaryTree::Node* Search(const int, Node*);
	    BinaryTree::Node* GetRoot();
	    BinaryTree::Node* FindMax(Node*);
	    BinaryTree::Node* FindMin(Node*);
	    BinaryTree::Node* Successor(const int);
        void PreOrder(Node*, void Print(const int)) const;
        void UpdateInfo();
        void CountSpecialKeys(Node*, std::vector<int>&);
	private:
	    BinaryTree::Node* CloneNode(Node*);
	    void Erase(Node*);
        int Height(Node*);
        void PreOrderHeights(Node*);
        void UpdateSubtreesSizes(Node*);
        void UpdateSubtreesHeights(Node*);
};

BinaryTree::BinaryTree() {
    Root = NULL;
}

BinaryTree::~BinaryTree() {
    Erase(Root);
}

void BinaryTree::Erase(Node* t) {
    if (t != NULL) {
        Erase(t->LSon);
        Erase(t->RSon);
        delete t;
    }
}

BinaryTree::Node* BinaryTree::Insert(Node* t, const int &value) {
	if (Search(value, t)) {
        return NULL;
    }

	if (t == NULL) {
		t = new Node;
		t->Info = value;
		t->LSon = t->RSon = NULL;
	} else if (value < t->Info) {
			t->LSon = Insert(t->LSon, value);
    } else {
		t->RSon = Insert(t->RSon, value);
    }

	if (Root == NULL) {
		Root = t;
    }
	return t;
}

BinaryTree::Node* BinaryTree::EraseByValue(Node* t, int &value) {
	if (t == NULL) {
		return t;
    } else if (value < t->Info) {
			t->LSon = EraseByValue(t->LSon, value);
        } else if (value > t->Info) {
				t->RSon = EraseByValue(t->RSon, value);
            } else if (t->LSon == NULL && t->RSon == NULL) {
					if (t == Root) {
						Root = NULL;
                    } else {
						delete t;
						t = NULL;
					}
				} else if (t->LSon == NULL) {
						Node* temp = t;
						t = t->RSon;
						if (temp == Root) {
							Root = t;
                        }
						delete temp;
					} else if (t->RSon == NULL) {
							Node* temp = t;
							t = t->LSon;
							if (temp == Root) {
								Root = t;
                            }
							delete temp;
					} else {
							Node* temp = FindMin(t->RSon);
							t->Info = temp->Info;
							t->RSon = EraseByValue(t->RSon, temp->Info);
					}
	return t;
}

BinaryTree::Node* BinaryTree::Search(const int value, Node* t) {
    if (t == NULL || value == t->Info) {
        return t;
    }

    if (value < t->Info) {
		return Search(value, t->LSon);
	} else {
			return Search(value, t->RSon);
	}
}

void BinaryTree::PreOrder(Node* t, void Print(const int)) const {
	if (Root == NULL) {
		throw "null\n";
    }

	if (t == NULL) {
		return;
    }

    Print(t->Info);
	PreOrder(t->LSon, Print);
	PreOrder(t->RSon, Print);
}

void BinaryTree::Node::PreOrderCount(Node* node, std::vector<int>& v) const {
    if (node == NULL) {
        return;
    }

    v.push_back(node->Info);
    PreOrderCount(node->LSon, v);
    PreOrderCount(node->RSon, v);
}

void BinaryTree::Node::CountSubtrees(Node* node) {
    if (node->LSon != NULL) {
        std::vector<int> v;
        Node::PreOrderCount(node->LSon, v);
        node->LSize = v.size();
    }

    if (node->RSon != NULL) {
        std::vector<int> v;
        Node::PreOrderCount(node->RSon, v);
        node->RSize = v.size();
    }
}

int BinaryTree::Height(Node* node) {
    if (node == NULL) {
        return 0;
    }

    if (node->LSon != NULL) {
        node->LHeight = Height(node->LSon);
    } else {
        node->LHeight = -1;
    }

    if (node->RSon != NULL) {
        node->RHeight = Height(node->RSon);
    } else {
        node->RHeight = -1;
    }

    return std::max(node->LHeight, node->RHeight) + 1;
}

void BinaryTree::PreOrderHeights(Node* node) {
	if (node == NULL) {
		return;
    }

    node->LHeight = Height(node->LSon);
    node->RHeight = Height(node->RSon);
	PreOrderHeights(node->LSon);
	PreOrderHeights(node->RSon);
}

void BinaryTree::UpdateSubtreesHeights(Node* node) {
    if (node == NULL) {
        return;
    }

    PreOrderHeights(node);
    UpdateSubtreesHeights(node->LSon);
    UpdateSubtreesHeights(node->RSon);
}

void BinaryTree::UpdateSubtreesSizes(Node* node) {
    if (node == NULL) {
        return;
    }

    node->CountSubtrees(node);
    UpdateSubtreesSizes(node->LSon);
    UpdateSubtreesSizes(node->RSon);
}

void BinaryTree::UpdateInfo() {
    UpdateSubtreesSizes(Root);
    UpdateSubtreesHeights(Root);
}

void BinaryTree::CountSpecialKeys(Node* node, std::vector<int>& v) {
    if (node == NULL) {
        return;
    }

    if (node->LSon != NULL && node->RSon != NULL) {
        if (node->LHeight == node->RHeight && node->LSize != node->RSize) {
            v.push_back(node->Info);
        }
    }
    CountSpecialKeys(node->LSon, v);
    CountSpecialKeys(node->RSon, v);
}

BinaryTree::Node* BinaryTree::GetRoot() {
    return Root;
}

BinaryTree::Node* BinaryTree::FindMax(Node* x) {
    if (x == NULL || x->RSon == NULL) {
        return x;
    }
    return FindMax(x->RSon);
}

BinaryTree::Node* BinaryTree::FindMin(Node* x) {
    if (x == NULL || x->LSon == NULL) {
        return x;
    }
    return FindMin(x->LSon);
}

BinaryTree::Node* BinaryTree::Successor(const int value) {
    Node* Current = Root;
    Node* SuccessorEl = NULL;
    while (Current != NULL) {
        if (value < Current->Info) {
            SuccessorEl = Current;
            Current = Current->LSon;
        } else {
            Current = Current->RSon;
        }
    }
    return SuccessorEl;
}

std::ifstream in("in.txt");
std::ofstream out("out.txt");

void Print(const int v) {
    out << v << std::endl;
}

int main() {
    int key;

    BinaryTree* bst = new BinaryTree();
    while (in >> key) {
        bst->Insert(bst->GetRoot(), key);
    }
    in.close();

    bst->UpdateInfo();
    std::vector<int> v_keys;
    bst->CountSpecialKeys(bst->GetRoot(), v_keys);
    if (v_keys.size() % 2 != 0) {
        std::sort(v_keys.begin(), v_keys.end());
        int key_to_remove = v_keys[v_keys.size() / 2];
        bst->EraseByValue(bst->GetRoot(), key_to_remove);
    }

    bst->PreOrder(bst->GetRoot(), Print);
    out.close();

    delete bst;
    return 0;
}

